tests/regression/test_mesh_vertexonly.py

```diff
@@ -0,0 +1,333 @@
+    assert vm.num_cells() == len(inputvertexcoords[in_bounds]) == vm.cell_set.size
+    assert vm.num_facets() == 0
+    assert vm.num_faces() == vm.num_entities(2) == 0
+    assert vm.num_edges() == vm.num_entities(1) == 0
+    assert vm.num_vertices() == vm.num_entities(0) == vm.num_cells()
+    # Correct parent cell numbers
+    stored_vertex_coords = np.copy(vm._swarm.getField("DMSwarmPIC_coor")).reshape((vm.num_cells(), gdim))
+    vm._swarm.restoreField("DMSwarmPIC_coor")
+    stored_parent_cell_nums = np.copy(vm._swarm.getField("parentcellnum"))
+    vm._swarm.restoreField("parentcellnum")
+    assert len(stored_vertex_coords) == len(stored_parent_cell_nums)
+    for i in range(len(stored_vertex_coords)):
+        assert m.locate_cell(stored_vertex_coords[i]) == stored_parent_cell_nums[i]
+
+
+@pytest.mark.parametrize("parentmesh", parentmeshes)
+def test_generate_cell_midpoints(parentmesh):
+    """
+    Generate cell midpoints for mesh m and check they lie in the correct cells
+    """
+    inputcoords, inputcoordslocal = cell_midpoints(parentmesh)
+    vm = VertexOnlyMesh(parentmesh, inputcoords)
+    # Midpoints located in correct cells of parent mesh
+    V = VectorFunctionSpace(parentmesh, "DG", 0)
+    f = Function(V).interpolate(parentmesh.coordinates)
+    # Check size of biggest len(vm.coordinates.dat.data_ro) so
+    # locate_cell can be called on every processor
+    max_len = MPI.COMM_WORLD.allreduce(len(vm.coordinates.dat.data_ro), op=MPI.SUM)
+    out_of_mesh_point = np.empty(shape=(1, parentmesh.geometric_dimension()))
+    out_of_mesh_point.fill(np.inf)
+    for i in range(max_len):
+        if i < len(vm.coordinates.dat.data_ro):
+            cell_num = parentmesh.locate_cell(vm.coordinates.dat.data_ro[i])
+        else:
+            cell_num = parentmesh.locate_cell(out_of_mesh_point)  # should return None
+        if cell_num is not None:
+            assert all(f.dat.data_ro[cell_num] == vm.coordinates.dat.data_ro[i])
+
+
+@pytest.mark.parallel
+@pytest.mark.parametrize("parentmesh", parentmeshes)
+def test_generate_cell_midpoints_parallel(parentmesh):
+    test_generate_cell_midpoints(parentmesh)
+
+
+@pytest.mark.parametrize("parentmesh", parentmeshes)
+@pytest.mark.parametrize("n", ncoords)
+def test_generate_random(parentmesh, n):
+    gdim = parentmesh.geometric_dimension()
+    inputcoords = random_coords(n, gdim)
+    vm = VertexOnlyMesh(parentmesh, inputcoords)
+    verify_vertexonly_mesh(parentmesh, vm, inputcoords, gdim)
+
+
+@pytest.mark.parallel
+@pytest.mark.parametrize("parentmesh", parentmeshes)
+@pytest.mark.parametrize("n", ncoords)
+def test_generate_random_parallel(parentmesh, n):
+    test_generate_random(parentmesh, n)
+
+
+@pytest.mark.parametrize("parentmesh", parentmeshes)
+@pytest.mark.xfail(raises=NotImplementedError)
+def test_extrude(parentmesh):
+    inputcoords, inputcoordslocal = cell_midpoints(parentmesh)
+    vm = VertexOnlyMesh(parentmesh, inputcoords)
+    ExtrudedMesh(vm, 1)
```

Comment:
Prefer

```
with pytest.raises(NotImplementedError):
   ExtrudedMesh(vm, 1)
```

over xfailing with raises.

Reply:
Surely I want an xfail because I haven't implemented, rather than a pass? It's a test I expect to fail after all

Reply:
the mechanism I propose checks that `ExtrudedMesh(vm, 1)` raises `NotImplementedError`, and will therefore fail as soon as you implement it.

Reply:
Yes I follow, but everything else that isn't implemented is marked as an xfail - why should this be different and why is it better? Is it because I can attach it to the test generation tests and can then reduce the test load?

Reply:
(I originally did what you said for all the `NotImplementedError` cases, but David said it they should really xfail so I don't know who to go with)

Reply:
I don't really care that much. I guess it's a philosophical thing of "either the test asserts that ExtrudedMesh is NotImplemented" or "the test runs, but it is marked as known to fail"

Reply:
I shall leave it be for now then.